Modules/signalmodule.c

```diff
@@ -1737,20 +1726,25 @@ _PyErr_CheckSignals(void)
 }
 
 
-/* Simulate the effect of a signal.SIGINT signal arriving. The next time
-   PyErr_CheckSignals is called,  the Python SIGINT signal handler will be
-   raised.
+/* Simulate the effect of a signal arriving. The next time PyErr_CheckSignals
+   is called,  the corresponding Python signal handler will be raised.
 
-   Missing signal handler for the SIGINT signal is silently ignored. */
+   Missing signal handler for the given signal number is silently ignored. */
 void
-PyErr_SetInterrupt(void)
+PyErr_SetInterruptEx(int signum)
 {
-    if ((Handlers[SIGINT].func != IgnoreHandler) &&
-        (Handlers[SIGINT].func != DefaultHandler)) {
-        trip_signal(SIGINT);
+    if ((Handlers[signum].func != IgnoreHandler) &&
+        (Handlers[signum].func != DefaultHandler)) {
+        trip_signal(signum);
```

Comment:
For the record, it's not obvious that the `if` condition is async-signal-safe. On a reasonable CPU, it probably is (and the condition is probably not necessary anyway, since `trip_signal` can be called for _any_ signal).

However, there is another issue: the `Handlers` array may be modified by another thread after the check and before the `trip_signal` call. Therefore `PyErr_CheckSignals` can later be called and try to execute e.g. `IgnoreHandler` (which is an integer object).

Reply:
> However, there is another issue: the `Handlers` array may be modified by another thread after the check and before the `trip_signal` call

Hummm...  I was under the assumption that other threads cannot register handlers

Reply:
I will try to make a reproducer.

Reply:
Ok, created https://bugs.python.org/issue43406

Reply:
Thanks, I think I understand now the situation you are mentioning.